src/preprocessing/particle_packing/system.jl

```diff
@@ -0,0 +1,287 @@
+                           [`SignedDistanceField`](@ref) of `boundary` with:
+                              - `use_for_boundary_packing=true`
+                              - `max_signed_distance=boundary_thickness`
+- `signed_distance_field`: To constrain particles onto the surface, the information about
+                           the signed distance from a particle to a face is required.
+                           The precalculated signed distances will be interpolated
+                           to each particle during the packing procedure.
+- `smoothing_kernel`:      Smoothing kernel to be used for this system.
+                           See [Smoothing Kernels](@ref smoothing_kernel).
+- `smoothing_length`:      Smoothing length to be used for this system.
+                           See [Smoothing Kernels](@ref smoothing_kernel).
+"""
+struct ParticlePackingSystem{NDIMS, ELTYPE <: Real, IC, B, K,
+                             S, N} <: FluidSystem{NDIMS, IC}
+    initial_condition     :: IC
+    boundary              :: B
+    smoothing_kernel      :: K
+    smoothing_length      :: ELTYPE
+    background_pressure   :: ELTYPE
+    tlsph                 :: Bool
+    signed_distance_field :: S
+    is_boundary           :: Bool
+    shift_condition       :: ELTYPE
+    neighborhood_search   :: N
+    signed_distances      :: Vector{ELTYPE} # Only for visualization
+    buffer                :: Nothing
+    update_callback_used  :: Ref{Bool}
+
+    function ParticlePackingSystem(shape::InitialCondition;
+                                   boundary::Union{Polygon, TriangleMesh},
+                                   signed_distance_field::SignedDistanceField,
+                                   smoothing_kernel=SchoenbergCubicSplineKernel{ndims(shape)}(),
+                                   smoothing_length=1.2 * shape.particle_spacing,
+                                   is_boundary=false, boundary_compress_factor=1.0,
+                                   neighborhood_search=GridNeighborhoodSearch{ndims(shape)}(),
+                                   background_pressure, tlsph=true)
+        NDIMS = ndims(shape)
+        ELTYPE = eltype(shape)
+
+        if ndims(smoothing_kernel) != NDIMS
+            throw(ArgumentError("smoothing kernel dimensionality must be $NDIMS for a $(NDIMS)D problem"))
+        end
+
+        # TODO: Let `Semidiscretization` handle this?
+        # Create neighborhood search
+        nhs_ = isnothing(neighborhood_search) ? TrivialNeighborhoodSearch{NDIMS}() :
+               neighborhood_search
+        nhs = copy_neighborhood_search(nhs_,
+                                       compact_support(smoothing_kernel, smoothing_length),
+                                       length(signed_distance_field.positions))
+
+        # Initialize neighborhood search with signed distances
+        PointNeighbors.initialize_grid!(nhs, stack(signed_distance_field.positions))
```

Comment:
@efaulhaber should I use a full grid nhs here? The positions of the `signed_distance_field` are of course static. Or what is the best choice here?

Reply:
Yes, by default the nhs should be using a `FullGridCellList`. You need to set this in the kwarg.
This cell list is faster and GPU-compatible. Only downsides are 1. one needs to specify a bounding box and 2. the domain is constrained to this bounding box. Here you know the bounding box and particles should never leave it, right?

Reply:
> Here you know the bounding box and particles should never leave it, right?

Yes, exactly. I also wondering if I use a precomputed NHS here. What do you think?